### PR TITLE
reservation: fix user 

### PR DESCRIPTION
**Pull request by alicefr**
**What this PR does / why we need it**:
Fix the user for the pr-helper container and fixes the directory permission on device creation.
This PR contains should be merged after #10186.
See [comment](https://github.com/kubevirt/kubevirt/pull/10186#issuecomment-1665222614) about the directory permission and [comment](https://github.com/kubevirt/kubevirt/pull/10186#issuecomment-1665289361) for the user.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-handler/device-manager/socket_device.go [hunk 1]
```diff
@@ -211,6 +212,14 @@ func (dpi *SocketDevicePlugin) Allocate(ctx context.Context, r *pluginapi.Alloca
 	response := pluginapi.AllocateResponse{}
 	containerResponse := new(pluginapi.ContainerAllocateResponse)
 
+	prSockDir, err := safepath.JoinAndResolveWithRelativeRoot("/", dpi.socketDir)
+	if err != nil {
+		panic(err)
```
Comment (vasiliy-ul) [inline pkg/virt-handler/device-manager/socket_device.go:217]:
```suggestion
		return nil, fmt.Errorf("error opening the socket directory %s: %v", dpi.socketDir, err)
```

File: pkg/virt-handler/device-manager/socket_device.go [hunk 2]
```diff
@@ -211,6 +212,14 @@ func (dpi *SocketDevicePlugin) Allocate(ctx context.Context, r *pluginapi.Alloca
...
+	}
+	err = safepath.ChownAtNoFollow(prSockDir, util.NonRootUID, util.NonRootUID)
+	if err != nil {
+		panic(err)
```
Comment (vasiliy-ul) [inline pkg/virt-handler/device-manager/socket_device.go:221]:
```suggestion
		return nil, fmt.Errorf("error setting permissions for the directory %s: %v", dpi.socketDir, err)
```

File: pkg/virt-handler/device-manager/socket_device.go (hunk 1 above)
Comment (alicefr) [reply pkg/virt-handler/device-manager/socket_device.go]: Done

File: pkg/virt-handler/device-manager/socket_device.go (hunk 2 above)
Comment (alicefr) [reply pkg/virt-handler/device-manager/socket_device.go]: Done!

**Comment by vasiliy-ul** [comment]
Hm... Even with root user, I still get:
```console
$ k exec -ti -n kubevirt virt-handler-gdxxh -c pr-helper -- bash
bash-5.1# /usr/bin/qemu-pr-helper -k /var/run/kubevirt/daemons/pr/pr-helper2.sock
^Cqemu-pr-helper: Failed to unlink socket /var/run/kubevirt/daemons/pr/pr-helper2.sock: Permission denied
bash-5.1# 
bash-5.1# ls -la /var/run/kubevirt/daemons/pr/
total 0
drwxr-xr-x. 2  107  107 80 Aug  7 08:21 .
drwxr-xr-x. 3 root root 16 Aug  7 07:53 ..
srwxr-xr-x. 1 root root  0 Aug  7 07:53 pr-helper.sock
srwxrwxrwx. 1 root root  0 Aug  7 08:21 pr-helper2.sock
```

File: pkg/virt-handler/device-manager/socket_device.go [hunk 3]
```diff
@@ -212,6 +212,14 @@ func (dpi *SocketDevicePlugin) Allocate(ctx context.Context, r *pluginapi.Alloca
...
 	containerResponse := new(pluginapi.ContainerAllocateResponse)
 
+	prSockDir, err := safepath.JoinAndResolveWithRelativeRoot("/", dpi.socketDir)
+	if err != nil {
+		return nil, fmt.Errorf("error opening the directory %s: %v", dpi.socketDir, err)
+	}
+	err = safepath.ChownAtNoFollow(prSockDir, util.NonRootUID, util.NonRootUID)
+	if err != nil {
+		return nil, fmt.Errorf("error setting directory ownership %s: %v", dpi.socketDir, err)
+	}
```
Comment (vasiliy-ul) [inline pkg/virt-handler/device-manager/socket_device.go:222]: I have a feeling that we may not need to chown the directory at all when running pr-helper as root. It should have `0755` by default, thus it's world-readable. Will do a quick test...

Comment (vasiliy-ul) [reply pkg/virt-handler/device-manager/socket_device.go:222]: Yep, no error on unlink if the directory is not chown'ed:
```
$ k exec -ti -n kubevirt virt-handler-wxf5m -c pr-helper -- bash
bash-5.1# ls -la /var/run/kubevirt/daemons/pr/
total 0
drwxr-xr-x. 2 root root 60 Aug  7 08:44 .
drwxr-xr-x. 3 root root 16 Aug  7 08:44 ..
srwxr-xr-x. 1 root root  0 Aug  7 08:44 pr-helper.sock
bash-5.1# 
bash-5.1# /usr/bin/qemu-pr-helper -k /var/run/kubevirt/daemons/pr/pr-helper2.sock
^Cbash-5.1# 
bash-5.1# 
bash-5.1# ls -la /var/run/kubevirt/daemons/pr/
total 0
drwxr-xr-x. 2 root root 60 Aug  7 08:44 .
drwxr-xr-x. 3 root root 16 Aug  7 08:44 ..
srwxr-xr-x. 1 root root  0 Aug  7 08:44 pr-helper.sock
```

Comment (vasiliy-ul) [reply pkg/virt-handler/device-manager/socket_device.go:222]:
```
Ran 3 of 1520 Specs in 757.019 seconds
SUCCESS! -- 3 Passed | 0 Failed | 2 Pending | 1515 Skipped
```

**Review by vasiliy-ul — APPROVED** [review]
/approve

nit: probably the commit message needs an update
```
reservation: chown and relabel pr dir on device creation
```
This does not reflect anymore what is actually going on in the code. I would squash the two commits in one, since they sort of depend on each other. But up to you.

**Comment by vasiliy-ul** [comment]
> I'm not sure what is happening, but don't get the unlink error message you reported here.

Probably you need to run a VM with reservation first, so the device plugin allocates the resource and chown's the directory to 107:107.

**Comment by alicefr** [comment]
> > I'm not sure what is happening, but don't get the unlink error message you reported here.
> 
> Probably you need to run a VM with reservation first, so the device plugin allocates the resource and chown's the directory to 107:107.

Ok, the issue happen when the directory `/var/run/kubevirt/daemons/pr/` is owned by users 107, if it belongs to root then the issue doesn't occurs and the pr-helper successfully unlinks the socket

**Comment by vasiliy-ul** [comment]
Correct, that is what I observed as well.

**Review by vasiliy-ul — APPROVED** [review]
/retest-required

**Pull request merged**